### PR TITLE
add:batch details for a user

### PR DESCRIPTION
**Pull request by nishihere19**
Signed-off-by: nishihere19 <nishihere19@gmail.com>

## Proposed Changes

  - Added new model
  - Added relevant APIs to fetch the batches from db, update the profile, update the batch separately.

### CODE REVIEW

File: prisma/schema.prisma [hunk 1]
```diff
@@ -35,6 +35,7 @@ model ResourceOwner {
   Code           Code[]
   authorizedApps AuthorisedApps[]
   Token          Token[]
+  batch          String           @default("NA")
```
Comment (siva2204) [inline prisma/schema.prisma:38]: can we change this to Enum like `BTech First Year`, `BTech Second Year`, if we need this data for proper things

or else we can go with the old one, @default can be `NULL`

File: prisma/schema.prisma [hunk 2]
```diff
@@ -105,3 +106,8 @@ model Codechallenge {
   createdAt                 DateTime      @default(now())
   expireAt                  DateTime
 }
+
+model Batch{
+  id          Int           @id @default(autoincrement())
+  batch       String        @unique
+}
```
Comment (siva2204) [inline prisma/schema.prisma:113]: oh you're using this to store all the batch and year, then its better to create new table to store batch ? (many to many relations)

File: src/controllers/authContoller.ts [hunk 3]
```diff
@@ -424,7 +435,7 @@ export const updateProfile = async (
     return res.status(400).json({ errors: errors.array() });
   }
 
-  const { name, phoneNumber, gender } = req.body;
+  const { name, phoneNumber, gender, batch } = req.body;
```
Comment (siva2204) [inline src/controllers/authContoller.ts:438]:
```   
const userBatch = await prisma.batch.findUnique({
      where: {
        batch: batch.toString()
      }
    });

```

where is this check, we can have dropdown in frontend ?

File: src/controllers/authContoller.ts [hunk 4]
```diff
@@ -445,3 +457,48 @@ export const updateProfile = async (
...
+): Promise<unknown> => {
+  const errors = validationResult(req);
+  if (!errors.isEmpty()) {
+    return res.status(400).json({ errors: errors.array() });
+  }
+
+  const { batch } = req.body;
+  const user: any = req.user;
+  try {
+    // create user
```
Comment (siva2204) [inline src/controllers/authContoller.ts:489]: ??

File: src/controllers/authContoller.ts [hunk 5]
```diff
@@ -445,3 +457,48 @@ export const updateProfile = async (
...
+        batch: true
+      }
+    });
+    return res.status(200).json(batches);
+  } catch (error) {
+    return res.status(500).json({ message: 'Internal server error' });
+  }
+};
+
+export const updateBatch = async (
```
Comment (siva2204) [inline src/controllers/authContoller.ts:477]: we can reuse the update profile route itself right ?
cuz ppl have to update their gender too, we have assigned `Male` for all the users

File: src/controllers/dashboardController.ts [hunk 6]
```diff
@@ -15,6 +15,7 @@ export const userInfo = async (
       select: {
         name: true,
         phoneNumber: true,
+        batch: true,
```
Comment (siva2204) [inline src/controllers/dashboardController.ts:18]: pls fix that `undefined` in the dashboard this time

File: src/routes/auth.ts [hunk 7]
```diff
@@ -43,6 +45,12 @@ router.post(
   updateProfile
 );
 
+// get batches route
+router.get('/batches', getBatches);
+
+// update batch route
+router.post('/updateBatch', isAuthenticated, updateBatch);
```
Comment (siva2204) [inline src/routes/auth.ts:52]: we can use `editProfile` route for this yeah ?

**Review by siva2204 — CHANGES_REQUESTED** [review]
where are you checking if the user has updated their profile ?

have to check during user authorization request.

File: src/controllers/authContoller.ts (hunk 3 above)
Comment (nishihere19) [reply src/controllers/authContoller.ts:438]: we already have it in frontend.

File: prisma/schema.prisma (hunk 2 above)
Comment (nishihere19) [reply prisma/schema.prisma:113]: Not needed imo, this is just to store the available batches from which the user can choose from.

File: src/controllers/dashboardController.ts (hunk 6 above)
Comment (nishihere19) [reply src/controllers/dashboardController.ts:18]: That has to be done in frontend, not backend.

File: prisma/schema.prisma (hunk 2 above)
Comment (siva2204) [reply prisma/schema.prisma:113]: storing same string/data(s) for each user is not best practice, using many-many relation will be apt here.
but anything is fine

File: src/controllers/authContoller.ts (hunk 3 above)
Comment (siva2204) [reply src/controllers/authContoller.ts:438]: browser is not only the client will be expecting :)

File: src/controllers/dashboardController.ts (hunk 6 above)
Comment (siva2204) [reply src/controllers/dashboardController.ts:18]: just reminding you for the 100th time :p

**Review by siva2204 — CHANGES_REQUESTED** [review]

File: src/controllers/authContoller.ts [hunk 8]
```diff
@@ -248,24 +249,32 @@ export const register = async (
...
 
-  const { name, password, phoneNumber, gender } = req.body;
+  const { name, password, phoneNumber, gender, batch } = req.body;
   try {
     // generate salt to hash password
     const salt = await bcrypt.genSalt(10);
     const hashPassword = await bcrypt.hash(password, salt);
 
+    //check if batch is valid
+    const allBatches = await getAllBatches();
```
Comment (siva2204) [inline src/controllers/authContoller.ts:259]: remove `await`, getAllBatches is not a async function

File: src/controllers/authContoller.ts [hunk 9]
```diff
@@ -248,24 +249,32 @@ export const register = async (
...
         password: hashPassword,
         phoneNumber: phoneNumber,
-        gender: gender
+        gender: gender,
+        batch: batch.toString()
       }
     });
     return res.status(200).json({ message: 'Registration successful' });
   } catch (error) {
+    // console.log(error);
```
Comment (siva2204) [inline src/controllers/authContoller.ts:277]: remove this

File: src/controllers/authContoller.ts [hunk 10]
```diff
@@ -424,18 +433,25 @@ export const updateProfile = async (
     return res.status(400).json({ errors: errors.array() });
   }
 
-  const { name, phoneNumber, gender } = req.body;
+  const { name, phoneNumber, gender, batch } = req.body;
   const user: any = req.user;
   try {
-    // create user
+    //check if batch is valid
+    const allBatches = await getAllBatches();
```
Comment (siva2204) [inline src/controllers/authContoller.ts:440]: remove `await`, getAllBatches is not a async function

File: src/controllers/clientController.ts [hunk 11]
```diff
@@ -152,7 +152,7 @@ export const deleteClient = async (
 
     return res.status(200).json({ message: 'Client deleted successfully.' });
   } catch (error) {
-    console.log(error);
+    // console.log(error);
```
Comment (siva2204) [inline src/controllers/clientController.ts:155]: remove this

**Review by siva2204 — CHANGES_REQUESTED** [review]
check if user has updated profile [here](https://github.com/delta/DAuth-Backend/blob/dev/src/controllers/oauthContoller.ts#L31), if not redirect them to update profile and redirect back here.

having this check in frontend is very bad and not a good practise.

File: src/utils/utils.ts [hunk 12]
```diff
@@ -92,3 +92,32 @@ export const generateForgotPasswordToken = (
   });
   return token;
 };
+
+export const getAllBatches = () => {
```
Comment (siva2204) [inline src/utils/utils.ts:96]: suggestion : we can have this array as global variable ?

Comment (nishihere19) [reply src/utils/utils.ts:96]: does not make any difference.

**Pull request closed without merging**